### DISCUSSION TITLE
Redirecting output from console to file, disabling color output

### DISCUSSION THREAD

**Discussion by tsieger** [dodona-edu/dolos#972] [Ideas]
Hi,
 I would like to capture the output from console to a file, but I get the following error:
``` 
[error] _a.getColorDepth is not a function
```
(Perhaps, the code attempts to learn whether the console is able to display colors?)
Can I turn the color output off somehow, please?
Also, the same problem appears on a color console (`rxvt-unicode-256color`) when running dolos from the provided docker container. Weird.
I'm running Node.js v16.14.0 on Fedora linux.
thanks

**Comment by rien** [comment]
I've encountered this error before, but I was of the impression that his has been fixed already. Can you check if you're running the latest version of Dolos? (v2.0.1)

With `--format csv` you can have Dolos write its results to machine readable CSV files, that wil be more machine-readable compared to the formatted console output. Or, which information do you need from a Dolos analysis?

**Comment by tsieger** [comment]
I'm running:
```
Dolos v2.0.1
Node v16.14.0
Tree-sitter v0.20.1
```
Yes, I'm aware of `-f csv`, however, it lacks the cute side-by-side view of similar chunks. 
My use case is to present selected similarities in a simple text file without the need for running a local web server. (Also, the side-by-side "pair" view does not work for me in the web page occasionally - the page stucks "loading results".) 
Having the pair view in text form would be a plus, and this is why I asked for the possibility of simply redirecting the output from console to file. (Of course, colorful text output to e.g. html file would be much better ;-). ) Just rambling...

**Reply by rien** [reply]
Nice to hear that someone is actually using thee console view. Most of our users prefer the web view because of it's simplicity, that's the reason why that view also got a bit more attention and the console view a bit less.

> (Also, the side-by-side "pair" view does not work for me in the web page occasionally - the page stucks "loading results".)

Can you look for errors in the console if you encounter this issue again? We've recently reworked that page, so there are probably still a few bugs left unfortunately.

I've made an issue for your problem (#975) and look into a fix. You're probably correct that we need an extra check to see if the output is a terminal or not.

**Comment by tsieger** [comment]
Well, I can reproduce the problem of "indefinite loading" of the pair view easily, but there is no error reported on the console I run dolos from. Shall I enable some debugs to see some?

Regarding alternatives to the web pair view, would generating some text file (e.g. simple html or markdown) for each pair of files, upon extra user request, make sense for you? It would come handy IMHO.

**Reply by rien** [reply]
The errors should be printed to the developer console in your web browser.

The dolos program in your terminal doesn't actually do more than serving the resulting CSV-files. The visualisations in the browser are all performed by the Vue javascript framework.

**Reply by tsieger** [reply]
OK. Browser console shows:
```
Error: The module 'tree-sitter-cpp' could not be found. Try to install it using npm or yarn, but it may not be supported (yet).
    getLanguageModule http://localhost:3000/js/8521.52b88f39.js:7
    findLanguage http://localhost:3000/js/8521.52b88f39.js:7
    A http://localhost:3000/js/8521.52b88f39.js:13
    r http://localhost:3000/js/8521.52b88f39.js:13
```
I've installed `tree-sitter-cpp` and it's working (dolos can parse cpp sources). Perhaps the UI needs some extra installation steps?

**Reply by rien** [reply]
This is a known bug (#957), a possible fix is downgrading to dolos v2.0.0, but I'll try to create an earlier fix in the following days.

**Reply by rien** [reply]
This is now fixed, so updating to Dolos v2.0.2 should also work.

**Reply by tsieger** [reply]
Thanks, it works!

**Comment by tsieger** [comment]
BTW, I've come across https://stackoverflow.com/questions/27397865/how-to-write-stdout-to-file-with-colors. The solution of saving colored text using `ansi2html` does not work (for the very same problem of `[error] _a.getColorDepth is not a function`, but the trick with `script` works:
```
script -q -c "dolos run -l cpp  a.cpp b.cpp" out.txt
cat out.txt | ansi2html > out.html
```
Still, it would be nice to have the possibility of generating color text pair view directly, preferably directly saving colored fragment match for pairs of files with similarity above some threshold (possibly saving them along the dolos report itself)?

**Reply by rien** [reply]
The error you're getting is a bug which we still need to fix (#975), so in the future it will indeed be possible to save the results to files.

You can play around with some options to do some filtering already. For example `-S 0.6` should  only show pairs with a similarity > 60% and `-s 10` should only print fragments longer than 10 kgrams.

**Reply by rien** [reply]
We're currently focusing on the web UI and performance of the analysis itself. While we will be fixing bugs, we won't be adding features to the CLI in the coming months unfortunately.

If you're interested in adding this feature, we can discuss how a possible PR would look like if you want?
